pipeline/mutate/mutator_hydrator.go

```diff
@@ -107,68 +110,83 @@ func NewMutatorHydrator(c configuration.Provider, d mutatorHydratorDependencies)
 		// This is a best-practice value.
 		BufferItems: 64,
 	})
-	return &MutatorHydrator{c: c, d: d, client: httpx.NewResilientClientLatencyToleranceSmall(nil), hydrateCache: cache}
+	return &MutatorHydrator{c: c, d: d,
+		p: sync.Pool{
+			New: func() interface{} {
+				b := &bytes.Buffer{}
+				b.Grow(poolBufCapacityBytes)
+				return b
+			},
+		},
+		client: httpx.NewResilientClientLatencyToleranceSmall(nil), hydrateCache: cache}
 }
 
 func (a *MutatorHydrator) GetID() string {
 	return "hydrator"
 }
 
-func (a *MutatorHydrator) cacheKey(config *MutatorHydratorConfig, session string) string {
-	return fmt.Sprintf("%s|%x", config.Api.URL, md5.Sum([]byte(session)))
+// cacheKey creates a (composite) cache key.
+func (a *MutatorHydrator) cacheKey(keys ...string) string {
+	return fmt.Sprintf("%x", md5.Sum([]byte(strings.Join(keys, ""))))
 }
 
-func (a *MutatorHydrator) hydrateFromCache(config *MutatorHydratorConfig, session string) (*authn.AuthenticationSession, bool) {
-	if !config.Cache.Enabled {
-		return nil, false
-	}
-
-	item, found := a.hydrateCache.Get(a.cacheKey(config, session))
+func (a *MutatorHydrator) hydrateFromCache(key string) (*authn.AuthenticationSession, bool) {
+	item, found := a.hydrateCache.Get(key)
 	if !found {
 		return nil, false
 	}
-
 	return item.(*authn.AuthenticationSession).Copy(), true
 }
 
 func (a *MutatorHydrator) hydrateToCache(config *MutatorHydratorConfig, key string, session *authn.AuthenticationSession) {
-	if !config.Cache.Enabled {
-		return
-	}
-
-	if a.hydrateCache.SetWithTTL(a.cacheKey(config, key), session.Copy(), 0, config.Cache.ttl) {
+	if a.hydrateCache.SetWithTTL(key, session.Copy(), 0, config.Cache.ttl) {
 		a.d.Logger().Debug("Cache reject item")
 	}
 }
 
-func (a *MutatorHydrator) Mutate(r *http.Request, session *authn.AuthenticationSession, config json.RawMessage, _ pipeline.Rule) error {
+func (a *MutatorHydrator) Mutate(r *http.Request, session *authn.AuthenticationSession, config json.RawMessage, p pipeline.Rule) error {
 	cfg, err := a.Config(config)
 	if err != nil {
 		return err
 	}
 
-	var b bytes.Buffer
-	if err := json.NewEncoder(&b).Encode(session); err != nil {
-		return errors.WithStack(err)
-	}
+	sessionCacheKey := a.p.Get().(*bytes.Buffer)
+	defer func() {
+		sessionCacheKey.Reset()
+		a.p.Put(sessionCacheKey)
+	}()
 
-	encodedSession := b.String()
-	if cacheSession, ok := a.hydrateFromCache(cfg, encodedSession); ok {
-		*session = *cacheSession
-		return nil
+	err = json.NewEncoder(sessionCacheKey).Encode(session)
+	switch {
+	case err != nil:
+		return errors.WithStack(err)
+	case !cfg.Cache.Enabled:
+		break
+	case len(cfg.Cache.Key) > 0:
+		// Build a composite cache key with property from configuration.
+		if cacheSession, ok := a.hydrateFromCache(a.cacheKey(
```

Comment:
Key is a composite key based on configuration property + three distinct repeatable properties.

Reply:
I am not sure if that is enough! What about extra properties of the session, such as e.g. "scope" or "permissions"? These could all influence the hydrator response and could lead to eventual security vulnerabilities down the road. I think we need to take the full session in a serialized form if we want to be sure that the cache can actually be reused!

Reply:
If we take the full `AuthenticationSession` we're not really benefiting anything given the volatiltity of this structure - using `subject`, `rule id`, `hydrator URL` as part of the composition is about enabling flexibility while reducing chance of cache leakage. We also need to remember this is opt-in and not opt-out as part of configuration.

Reply:
Thank you. I am unsure though wether I can agree here. I think it points to a deeper issue though which is manifesting in this pr: It is not possible to define the cache key reliably. Actually, one would probably like to define the cache key themselves (e.g. subject + scope) to make the system more efficient. For example, the AuthSession might contain vital info such as a "permissions" array. Yet, it may also contain a counter or timestamp which is changing for most of the requests - invalidating the cache if included.

We are still not there yet with a new concept for Ory Oathkeeper, but I think this is a very interesting problem that could very well warrant a realignment on Ory Oathkeeper which would be to make access control at the reverse proxy as efficient and flexible as possible.

I think this too could benefit from JsonNet, as JsonNet is typable, lintable, and can produce errors (go templating fulfills none of these properties).

Unfortunately, for this PR in particular, I don't think the current implementation can be accepted because it still bears too many risks and this particular type of issue has already caused a CVE in Ory Oathkeeper which is why I am so hyper-sensible about this topic: https://github.com/ory/oathkeeper/security/advisories/GHSA-qvp4-rpmr-xwrr

Reply:
Thank you. Let us inverse the solution. If you feel a user can't define a cache key freely (given cause of concerns as mentioned) - what about allowing the user to define exclusion of request/response `headers`? If the whole `AuthenticationSession` is used as a cache key `headers` represents a volatile part. `extra` is defined as part of previous steps in the pipeline and can be more controlled. What is your thoughts?

Reply:
That's a brilliant idea! Exclusion is much safer (because explicit) :)